zombies/src/main/java/org/phantazm/zombies/map/shop/interactor/SlotMachineInteractor.java

```diff
@@ -0,0 +1,346 @@
+                    newComponents.add(MiniMessage.miniMessage().deserialize(formatString, tags));
+                }
+
+                updateHologram(newComponents);
+                return;
+            }
+
+            for (ShopInteractor interactor : timeoutExpiredInteractors) {
+                interactor.handleInteraction(rollInteraction);
+            }
+
+            reset();
+        }
+    }
+
+    private void reset() {
+        if (item != null) {
+            item.remove();
+            item = null;
+        }
+
+        hologram.clear();
+        rollInteraction = null;
+        rolling = false;
+        doneRolling = false;
+        rollFinishTime = 0L;
+        lastFrameTime = 0L;
+        currentFrameIndex = 0;
+        ticksUntilNextFrame = 0;
+    }
+
+    private void displayRollFrame(int index) {
+        SlotMachineFrame frame = frames.get(index % frames.size());
+
+        if (item != null) {
+            item.remove();
+        }
+
+        item = new Entity(EntityType.ITEM);
+        ItemEntityMeta meta = (ItemEntityMeta)item.getEntityMeta();
+        meta.setItem(frame.getVisual());
+
+        item.setInstance(shop.instance(), shop.center().add(0, data.itemOffset, 0));
+
+        TagResolver[] tags = getTagsForFrame(frame);
+
+        List<Component> newComponents = new ArrayList<>(frame.hologramFormats().size());
+        for (String formatString : frame.hologramFormats()) {
+            newComponents.add(MiniMessage.miniMessage().deserialize(formatString, tags));
+        }
+
+        updateHologram(newComponents);
+    }
+
+    private TagResolver[] getTagsForFrame(SlotMachineFrame frame, TagResolver... additionalTags) {
+        TagResolver rollingPlayerTag = Placeholder.component("rolling_player",
+                rollInteraction.player().module().getPlayerView().getDisplayNameIfPresent());
+
+        Component displayName = frame.getVisual().getDisplayName();
+        if (displayName == null) {
+            displayName = Component.text(frame.getVisual().material().key().value());
+        }
+
+        TagResolver itemName = Placeholder.component("item_name", displayName);
+
+        TagResolver[] tags = new TagResolver[additionalTags.length + 2];
+        tags[0] = rollingPlayerTag;
+        tags[1] = itemName;
+
+        System.arraycopy(additionalTags, 0, tags, 2, additionalTags.length);
+        return tags;
+    }
+
+    private void updateHologram(List<Component> newComponents) {
+        while (hologram.size() > newComponents.size()) {
+            hologram.remove(hologram.size() - 1);
+        }
+
+        for (int i = 0; i < newComponents.size(); i++) {
+            Component newComponent = newComponents.get(i);
+            Component oldComponent = i < hologram.size() ? hologram.get(i) : null;
+
+            if (!newComponent.equals(oldComponent)) {
+                hologram.set(i, newComponent);
+            }
+        }
+    }
+
+    public interface SlotMachineFrame {
```

Comment:
why do we need need an interface here for a primarily internal class?

Reply:
just to make it easier to expand ig

Reply:
it is mostly personal preference but I think it just needlessly adds complexity